File: proteus_orocos_obstaclelaser/src/AvoidObstacleLaser.cpp
#include <rtt/TaskContext.hpp>
#include <rtt/Port.hpp>
#include <geometry_msgs/Twist.h>
#include <ocl/Component.hpp>
#include <sensor_msgs/LaserScan.h>
#include <numeric>

using namespace std;
using namespace RTT;

class AvoidObstacleLaser : public RTT::TaskContext {
private:  
  InputPort<sensor_msgs::LaserScan> inport;
  OutputPort<geometry_msgs::Twist> outport;

public:
  AvoidObstacleLaser(const std::string& name):
    TaskContext(name),
    inport("laser_in"),
    outport("twist_out")
  {
    ports()->addPort(inport);
    ports()->addPort(outport);
  }
  ~AvoidObstacleLaser() {}
private:
  void updateHook() {
    sensor_msgs::LaserScan msg;
    if (NewData == inport.read(msg)) {
      geometry_msgs::Twist cmd;

      bool halt = false;
      for (int i = 75; i <= 105; i++) {
        if (msg.ranges[i] < 2.0) {
          halt = true;
          break;
        }
      }
      if (halt) {
        double midA, midB;
        // we go to the highest-range side scanned
        midA = std::accumulate(msg.ranges.begin(), msg.ranges.end()-90, 0);
        midB = std::accumulate(msg.ranges.begin()+90, msg.ranges.end(), 0);
        log(Info)<<"A:"<<midA<<", B:"<<midB<<endlog();
        if (midA > midB) {
          cmd.angular.z = -1;
        } else {
          cmd.angular.z = 1;
        }
      } else {
        cmd.linear.x = 1;
      }
      outport.write(cmd);
    }
  }
};

ORO_CREATE_COMPONENT(AvoidObstacleLaser)

File: proteus_java/src/proteus/AvoidObstacleLaser.java
package proteus;

import com.google.common.base.Preconditions;

import org.apache.commons.logging.Log;
import org.ros.DefaultNode;
import org.ros.MessageListener;
import org.ros.Node;
import org.ros.NodeConfiguration;
import org.ros.NodeMain;
import org.ros.Publisher;
import org.ros.Subscriber;

import org.ros.message.geometry_msgs.Twist;
import org.ros.message.sensor_msgs.LaserScan;

public class AvoidObstacleLaser implements NodeMain {

	private Node node;

	@Override
	public void main(NodeConfiguration configuration) {
		Preconditions.checkState(node == null);
		Preconditions.checkNotNull(configuration);
		try {
			node = new DefaultNode("laser_cmd", configuration);
			final Log log = node.getLog();
			log.info("init");
			final Publisher<Twist> publisher = 
				node.createPublisher("/ATRV/Motion_Controller", "geometry_msgs/Twist");
			node.createSubscriber("/ATRV/Sick", "sensor_msgs/LaserScan", 
				new MessageListener<LaserScan>() {
				@Override
				public void onNewMessage(LaserScan msg) {
					int i, mid = msg.ranges.length / 2;
					Twist cmd = new Twist();
					// halt if an object is less than 2m in a 30deg angle
					boolean halt = false;
					for (i = mid - 15; i <= mid + 15; i++) {
						if (msg.ranges[i] < 2.0) {
							halt = true;
							break;
						}
					}
					if (halt) {
						double midA = 0, midB = 0;
						// we go to the highest-range side scanned
						for (i = 0; i < mid; i++)
							midA += msg.ranges[i];
						for (i = mid; i < msg.ranges.length; i++)
							midB += msg.ranges[i];
						log.info("A:" + midA + ", B:" + midB);
						if (midA > midB) {
							cmd.angular.z = -1;
						} else {
							cmd.angular.z = 1;
						}
					} else {
						cmd.linear.x = 1;
					}
					publisher.publish(cmd);
				}
			});
		} catch (Exception e) {
			if (node != null) {
				node.getLog().fatal(e);
			} else {
				e.printStackTrace();
			}
		}
	}

	@Override
	public void shutdown() {
		node.shutdown();
		node = null;
	}

}

